Favicons for web UI

dandaka opened the pull request:
I have added favicons and changed index.html. But in my local olad I can't start webserver (Angular?) to start serving these files. It is serving all old files (like logo.png), but won't serve new once. Since all files are minified and there is no documentation, I thought it would be much easier for former developers to include these files.

@daveol please have a look

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,14 @@
+{
+  "version": 1,
+  "name": "Fontend Boilerplate",
```
Frontend?

peternewman replied:
Actually doesn't this want to be OLA

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,14 @@
+{
+  "version": 1,
+  "name": "Fontend Boilerplate",
+  "description": "Fontend Boilerplate",
```
Again

peternewman commented on olad/www/new/img/favicons/manifest.json:
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "Fontend Boilerplate",
```
Frontend?

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,14 @@
+    "512": "firefox_app_512x512.png"
+  },
+  "developer": {
+    "name": "Vlad Rafeev",
```
"Open Lighting Project" I think, assuming this shows up in FF etc

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,14 @@
+  },
+  "developer": {
+    "name": "Vlad Rafeev",
+    "url": "dandaka.com"
```
https://www.openlighting.org

peternewman commented:
Thanks for doing this @dandaka ! I remember looking into it a year or so ago and then getting distracted, but it looks a lot easier now.

dandaka commented:
I have fixed references to other projects. Please have a look.

dandaka commented:
Why doesn't my files served from webserver? Angular settings?

peternewman commented:
The webserver is part of olad, so you'll need to do a full autoreconf/configure/make, then run olad/olad and browse to port 9090. You can probably fudge this given you don't necessarily care about the actual OLA stuff by just serving up olad/www from a webserver.

You'll also need to add all the new files to olad/www/Makefile.mk or equivalent as appropriate, depending on their location.

peternewman commented on olad/www/new/index.html:
```diff
@@ -7,6 +7,39 @@
   <title>OLA</title>
   <link rel="stylesheet" href="/new/libs/bootstrap/css/bootstrap.min.css"/>
   <link rel="stylesheet" href="/new/css/style.min.css"/>
+  <link rel="icon" type="image/png" sizes="192x192" href="img/favicons/android-chrome-192x192.png">
```
I think judging on what we've done elsewhere in the file, these may need to be fully qualified with /new/ on the start. It would be good to be consistent anyway.

peternewman commented:
Re-reading your comment about olad, it sounds like you are running it up, did you also have the deb installed? If so you'll probably need to use the --http-data-dir option to serve it up from your WIP folder:
http://docs.openlighting.org/ola/man/man1/olad.1.html

There is some documentation for the new UI here:
https://github.com/OpenLightingProject/ola/blob/master/javascript/new-src/README.md

We've also got Grunt and all that magic, so you could add the relevant stuff in there to allow us to regenerate the favicons if we change our logo, it's probably worth chatting to @daveol about that:
https://github.com/OpenLightingProject/ola/tree/master/javascript/new-src

We've also got the RDM test server, if you fancy adding favicons to that too:
https://github.com/OpenLightingProject/ola/tree/master/tools/rdm

And the RDM website :) :
https://github.com/OpenLightingProject/rdm-app
http://rdm.openlighting.org/

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,10 @@
+{
+  "version": 1,
+  "name": "Ola",
```
OLA please

peternewman commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,10 @@
+{
+  "version": 1,
+  "name": "Ola",
+  "description": "Ola",
```
Open Lighting Architecture

peternewman commented on olad/www/new/img/favicons/manifest.json:
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "Ola",
```
OLA

peternewman commented:
Which original image file did you use BTW, was it from https://github.com/OpenLightingProject/logos ?

dandaka commented:
I've fixed descriptions. I need clarification on your comment "Can't we keep these, or is it not worth it?".

When you accept the pull request, I'll squash all this into a commit.

Yes, I have used original logo from your repo.

I was using `gulp` to generate these images. I think it will be too much for me now to set up a grunt task for possible future updates.

I can post to https://github.com/OpenLightingProject/logos image, that I have used to generate favicons. And all favicons. So anyone could fork and use in other repositories.

dandaka commented:
BTW, I have added @2x version of logo in new web UI. And fixed relative path according to your guide `/new/img/` (it was different). Tested favicons in two browsers, seems to be working.

I'm running a simple webserver (from grunt) as you advised.

peternewman commented:
Thanks for putting Twitter and Opengraph back.

You needn't bother squashing, I regularly commit stuff unsquashed, in some cases with far more commits.

Perhaps just add a note to https://github.com/OpenLightingProject/ola/blob/master/javascript/new-src/README.md mentioning which tools you used, and a link to which source image. Sorry, I misread/assumed gulp and grunt were compatible, or a sub or superset of each other. Would adding a gulp task just be a case of copying your boilerplate files? Do you fancy doing that instead?

Did you not just use this image to make the favicon, or was it not big enough:
https://github.com/OpenLightingProject/logos/blob/master/raster/OLA%20Logo%20O%20Only.png

I don't think we need every single favicon in the logos repo, as the number of those will just keep growing with new hardware, OSes and software, as long as we can generate them from an original.

Did you give up trying to serve them via olad then?

peternewman commented:
Regarding the 2x, that's great, I think that was what I was looking at ages ago, rather than favicons. Is there any reason you can't use our original logo-mini.png and this for the 2x? https://github.com/OpenLightingProject/logos/blob/master/web%20fitted/OLA-Logo-Fitted-22px%402x.png

It's just we've then got the source Photoshop file too.

peternewman commented:
The 2x stuff looks simple enough I might try updating our old UI too (unless you fancy doing the honours :) ) https://github.com/OpenLightingProject/ola/tree/master/olad/www .

nomis52 commented:
By default olad will serve the assets  in /usr/local/share/olad/www/ . You can point it somewhere else with the --http-data-dir flag.

nomis52 commented:
You also need to register these in olad/OladHTTPServer.cpp

peternewman commented:
Oops, thanks @nomis52 I forgot the need to register them. Perhaps we need a Travis check for that, files in the folder that aren't being served up.

peternewman commented:
Can you add ola-favicon.png to a Makefile please, to fix Travis.

dandaka commented:
@peternewman Should I put it under `dist_img_DATA`? I see no references to "new-src" folder in Makefile. Travis CI log looks huge. Can't find any guidelines about what and how to put in Makefile. And why. Need your help.

dandaka commented:
@nomis52 Can you help me with **olad/OladHTTPServer.cpp**? I can't test this code and have little idea, how it should be included.

peternewman commented:
@dandaka stick it in this Makefile:
https://github.com/OpenLightingProject/ola/blob/master/javascript/Makefile.mk

Essentially everything goes in a Makefile, or Travis will complain it won't ship.

Are you not in a position to compile the C++ then? Yes, you're basically there, add it the same as the other files already in there. We've got CONTENT_TYPE_CSS, CONTENT_TYPE_HTML and CONTENT_TYPE_JS depending on what you're serving up.

dandaka commented:
Did that.

*.json served with CONTENT_TYPE_HTML — ok?
*.ico served with CONTENT_TYPE_PNG — ok?

No, I'm not in a position to compile C++ :) I'm a frontend developer with little to none knowledge of CPP. I could spend too much time for this task.

daveol commented:
I wouldn't pick the html mimetype for json data, you can look [here](https://github.com/OpenLightingProject/ola/blob/master/olad/OladHTTPServer.cpp#L219) for mimetypes. You have three choices here:

 You make a new mimetype constant with type: application/json and name it like HTTPServer::CONTENT_TYPE_JSON 
example: common/http/HTTPServer.cpp
```cpp
const char HTTPServer::CONTENT_TYPE_JSON[] = "application/json";
```
example: olad/OladHTTPServer.cpp
```cpp
m_server.RegisterFile(
      "/new/img/favicons/yandex-browser-manifest.json",
      HTTPServer::CONTENT_TYPE_JSON);
```

You can use HTTPServer::CONTENT_TYPE_PLAIN (note might not work)
example: olad/OladHTTPServer.cpp
```cpp
m_server.RegisterFile(
      "/new/img/favicons/yandex-browser-manifest.json",
      HTTPServer::CONTENT_TYPE_PLAIN);
```

You can use HTTPServer::CONTENT_TYPE_OCT (which is basically the equivalent of "let the browser figure it out, because I have no idea")
example: olad/OladHTTPServer.cpp
```cpp
m_server.RegisterFile(
      "/new/img/favicons/yandex-browser-manifest.json",
      HTTPServer::CONTENT_TYPE_OCT);
```

For *.ico files you should probably just use the HTTPServer::CONTENT_TYPE_OCT

Also, all the files you have registered in olad/OladHTTPServer.cpp you have started with:
"olad/www/new" it should just be "/new"

Looks good so far @dandaka

daveol commented on olad/www/new/img/favicons/yandex-browser-manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "version": 1,
+  "api_version": 1,
+  "layout": {
+    "logo": "favicons/yandex-browser-50x50.png",
```
shouldn't this be ```/new/img/favicons/yandex-browser-50x50.png``` ?

daveol commented on olad/www/new/img/favicons/manifest.json:
```diff
@@ -0,0 +1,43 @@
+  "orientation": "portrait",
+  "icons": [
+    {
+      "src": "favicons/android-chrome-36x36.png",
```
shouldn't this be a full path? so like ```/new/img/favicons/android-chrome-36x36.png```

daveol commented on olad/www/new/img/favicons/manifest.webapp:
```diff
@@ -0,0 +1,10 @@
+  "name": "OLA",
+  "description": "Open Lighting Architecture",
+  "icons": {
+    "60": "firefox_app_60x60.png",
```
again, shouldn't this be a full path?

dandaka commented:
Should I add new content types to **common/http/HTTPServer.cpp** or to ** olad/OladHTTPServer.cpp**?

daveol commented:
[common/http/HTTPServer.cpp](/OpenlightingProject/ola/tree/master/common/http/HTTPServer.cpp) with the [rest of them](/OpenlightingProject/ola/tree/master/common/http/HTTPServer.cpp#L74)

dandaka commented:
Fixed content types and relative → absolute paths.

daveol commented on olad/OladHTTPServer.cpp:
```diff
@@ -219,6 +219,155 @@ OladHTTPServer::OladHTTPServer(ExportMap *export_map,
       "/new/img/logo-mini.png",
       HTTPServer::CONTENT_TYPE_PNG);
+  m_server.RegisterFile(
+      "olad/www/new/img/logo-mini@2x.png",
```
the path of the file is incorrect it should be ```/new/img/logo-mini@2x.png``` instead of ```olad/www/new/img/logo-mini@2x.png```. This is also true for the rest of the files you added

dandaka commented:
Sorry for that, fixed.

daveol commented on olad/www/Makefile.mk:
```diff
@@ -68,7 +68,56 @@ dist_img_DATA = \
-    olad/www/new/img/logo.png
+    olad/www/new/img/logo-mini@2x.png \
+    olad/www/new/img/logo.png \
+    olad/www/new/img/favicons/android-chrome-144x144.png \
```
if you want to keep the favicons directory you will have to specify a new variable at the top, something like: 
```make
favicondir = $(www_datadir)/new/img/favicons
```

and the at the begin of the files that you want to end up in the favicons dir you would have to specify:
```make
dist_favicon_DATA = \
```

otherwise autotools will just end up putting them in the ```/new/img``` directory

peternewman commented:
You're almost there, you need to add the new content types to include/ola/http/HTTPServer.h too, which should fix the remaining Travis issues.

peternewman commented:
Congratulations @dandaka you're changes are now passing all the tests (it just happens that flake8 has had pep8 updated, which has caused some existing code untouched by you to break). I'm just working on fixing that in https://github.com/OpenLightingProject/ola/pull/1014 so if you update when it's merged your tests should go green.

There are just a few other outstanding comments, e.g. https://github.com/OpenLightingProject/ola/pull/1008#issuecomment-170353996

Thanks again for all your work on this; it gets easier after the first bit! :)

dandaka commented:
> Is there any reason you can't use our original logo-mini.png and this for the 2x?

1. I have removed empty pixel from original `logo-mini.png`, no need for that
1. `OLA-Logo-Fitted-22px%402x.png` has empty pixels too, which is a bad practice IMO

Since a single pic is not a big deal, I have replaced them with better options. But you can always stick to old files, I just don't see any reason to do that.

If I have missed some other comments, please point me to them.

peternewman commented:
Can you resync please @dandaka and the checks should go green.

dandaka commented:
Did that.

peternewman commented:
Thanks @dandaka . This all looks great. Feel free to add yourself to here if you want a credit:
https://github.com/OpenLightingProject/ola/blob/master/AUTHORS

Understood about the empty pixels. Do you have Photoshop? It would be superb if https://github.com/OpenLightingProject/logos/tree/master/web%20fitted could be updated so what's available there matches what's in the Web UI.

The comment/request about favicons and 2x for the other interfaces would still be very welcome (as well as some other stuff in the existing UI that could be 2x, but I guess we'd need vector images for them).

@nomis52 or @daveol any final comments before we merge?

peternewman commented on olad/OladHTTPServer.cpp:
```diff
@@ -219,6 +219,155 @@ OladHTTPServer::OladHTTPServer(ExportMap *export_map,
+      HTTPServer::CONTENT_TYPE_PNG);
+  m_server.RegisterFile(
+      "/new/img/favicons/favicon.ico",
+      HTTPServer::CONTENT_TYPE_OCT);
```
https://en.wikipedia.org/wiki/ICO_%28file_format%29#MIME_type and http://stackoverflow.com/questions/13827325/correct-mime-type-for-favicon-ico suggest a MIME type of image/x-icon should be correct, would you mind changing to that please @dandaka ?

dandaka commented:
@peternewman Fixed issues with MIME.

I consider fixing other icons with a minor bug, which is not bothering right now.

You are welcome to open an issue for me in https://github.com/OpenLightingProject/logos/ project.

peternewman commented:
Thanks @dandaka . Issue raised here, I'll assign it when you accept the invitation:
https://github.com/OpenLightingProject/logos/issues/7

Understood about the other bits; I'm being cheeky really. Although TBH I didn't realise quite how easy it is to fix.

peternewman commented:
You'll need to resync again please @dandaka .

nomis52 commented:
LGTM. Lets merge this.

peternewman commented:
Great. @dandaka can you resync for a final time, then @ mention nomis52 and I and someone will hit the merge button.

dandaka commented:
@nomis52 @peternewman Ready to merge.

peternewman commented:
Thanks @dandaka !

dandaka commented:
Hurray, glad to contribute!

peternewman commented:
The question is what's next @dandaka ! :)
